### ISSUE TITLE
Possible fix for fitAll and centering behaving strange after "transform center of rotation" changes

### ISSUE THREAD

**Issue by mlimper** [x3dom/x3dom#572]
As reported by B. Lange on the x3dom-developers list:

Hello,

we noticed a strange behavior regarding rotation in examine mode when updating from x3dom 1.6.2 to 1.7.0. Where rotation worked fine for all our models with 1.6.2 it acted rather unintuitively with 1.7.0, but only for some models.

We found this to be related to these models having their <Viewpoint> inside a <Transform>, in which case examine mode started acting strange.

We believe that the changes in https://github.com/x3dom/x3dom/pull/505 "transform center of rotation" cause this behavior, when runtime.fitAll() is used. The same problem exists for the double click to center function. Andreas said in https://github.com/x3dom/x3dom/issues/504#issuecomment-110811186 that a new function setCenterOfRotationWC may be needed, this might be a use case.

We have patched Viewarea.js as follows:

```
diff --git a/src/Viewarea.js b/src/Viewarea.js
index 341a8ee..67e8964 100755
--- a/src/Viewarea.js
+++ b/src/Viewarea.js
@@ -1161,7 +1161,7 @@ x3dom.Viewarea.prototype.fit = function(min, max, updateCenterOfRotation)
      viewmat._23 = -viewDir.dot(eyePos);
  
      if (updateCenterOfRotation) {
-        viewpoint.setCenterOfRotation(center);
+        viewpoint.setCenterOfRotation(viewpoint.getCurrentTransform().inverse().multMatrixPnt(center));
      }
  
      if (x3dom.isa(viewpoint, x3dom.nodeTypes.OrthoViewpoint))
@@ -1633,7 +1633,7 @@ x3dom.Viewarea.prototype.onDoubleClick = function (x, y)
  
      var viewpoint = this._scene.getViewpoint();
  
-    viewpoint.setCenterOfRotation(this._pick);
+    viewpoint.setCenterOfRotation(viewpoint.getCurrentTransform().inverse().multMatrixPnt(this._pick));
      x3dom.debug.logInfo("New center of Rotation:  " + this._pick);
  
      var mat = this.getViewMatrix().inverse();
```

This works for us in that fitAll and centering by double click are behaving as expected again, both for models with transformed viewpoint and for those without.

**Comment by sressler** [comment]
Not exactly on this topic but relevant to Examine mode...Question...I'd like to create either an option to examine mode or another navigation mode perhaps called "spin" that will function like the old Cosmo Player such then when the user flicks the cursor with the mouse the object spins and remains spinning and also optionally slows down. Does anyone have any opinion on whether this should be a new navigation mode or an option to Examine?

**Comment by andreasplesch** [comment]
It would be probably necessary to have a simple example scene to show the problematic behaviour but this all makes sense if I understand above correctly:
- center is in world coordinates
- viewpoint.setCenterOfRotation() in viewarea receives world coordinates but now after #505 should get local coordinates since the viewpoint's cor is in local coordinates by definition.
- the fix transforms world coordinates to local coordinates and then delivers to  .setCenterOfRotation()

So I think this looks ok to me if this is what is going on. The question is if there other places which use setCenterOfRotation() and if there are if it would make sense to apply the fix inside the setCenterOfRotation() function (in viewpoint.js et al.)

Strictly speaking, I suggest it would be cleaner and in the spirit of the spec. to reserve setCenterOfRotation() (and getCenterOfRotation()) for local coordinates and introduce a new convenience function setCenterOfRotationWC() for world coordinates but ..
.. it is more convenient and x3dom backward compatible to patch setCenterOfRotation() to accept world coordinates, similar to getCenterOfRotation().

my 2c

**Comment by andreasplesch** [comment]
Hi Sandy, in my view a spinning option would be really nice and likely well received by everybody. I see it as a variation of the rotation around the cor in examine mode while dragging with the left mouse button. So perhaps CTRL-left click dragging could be used to invoke spinning in examine mode (if pressing the CTRL key is detectable) ? (Not sure how one button mice currently work with x3dom).
I have no real pointers on how to implement. There is an animateTo() (or so) function which would not directly apply but could perhaps be used as a template.

**Comment by sressler** [comment]
Hi Andreas...yes I like the idea of this being an extension to examine mode (at least for starters) ..I'll set up a separate issue to track this ...Sandy
